### PR TITLE
ocm: add new placement binding builder and unit tests

### PR DESCRIPTION
**Pull request by klaskosk**
Added `NewPlacementBindingBuilder` as well as methods to add a new `Subject` to the set and validate the `Subject` and `PlacementRef`. Tests for the new methods are included.

### CODE REVIEW

File: pkg/ocm/placementbinding.go [hunk 1]
```diff
@@ -180,6 +222,75 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
...
+// validatePlacementRef validates all the fields of the PlacementRef and returns an errorMsg based on the validation.
+// The errorMsg will be empty for valid Subjects.
+func validatePlacementRef(placementRef *policiesv1.PlacementSubject) string {
+	var errorMsg string
+
+	apiGroup := placementRef.APIGroup
+	if apiGroup != "apps.open-cluster-management.io" && apiGroup != "cluster.open-cluster-management.io" {
+		glog.V(100).Info("The APIGroup of the PlacementRef of the PlacementBinding is invalid")
+
+		errorMsg = "placement binding 'PlacementRef.APIGroup' must be a valid option"
```
Comment (sebrandon1) [inline pkg/ocm/placementbinding.go:249]:
```suggestion
		return "placement binding 'PlacementRef.APIGroup' must be a valid option"
```

File: pkg/ocm/placementbinding.go [hunk 2]
```diff
@@ -180,6 +222,75 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
...
+	}
+
+	kind := placementRef.Kind
+	if kind != "PlacementRule" && kind != "Placement" {
+		glog.V(100).Info("The Kind of the PlacementRef of the PlacementBinding is invalid")
+
+		errorMsg = "placement binding 'PlacementRef.Kind' must be a valid option"
```
Comment (sebrandon1) [inline pkg/ocm/placementbinding.go:256]:
```suggestion
		return "placement binding 'PlacementRef.Kind' must be a valid option"
```

File: pkg/ocm/placementbinding.go [hunk 3]
```diff
@@ -180,6 +222,75 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
...
+	}
+
+	if placementRef.Name == "" {
+		glog.V(100).Info("The Name of the PlacementRef of the PlacementBinding is empty")
+
+		errorMsg = "placement binding 'PlacementRef.Name' cannot be empty"
```
Comment (sebrandon1) [inline pkg/ocm/placementbinding.go:262]:
```suggestion
		return "placement binding 'PlacementRef.Name' cannot be empty"
```

File: pkg/ocm/placementbinding.go [hunk 4]
```diff
@@ -180,6 +222,75 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
...
+
+// validateSubject validates the fields of the Subject and returns an errorMsg based on the validation. The errorMsg
+// will be empty for valid Subjects.
+func validateSubject(subject *policiesv1.Subject) string {
+	var errorMsg string
+
+	if subject.APIGroup != "policy.open-cluster-management.io" {
+		glog.V(100).Info("The APIGroup of the PlacementBinding subject is invalid")
+
+		errorMsg = "placement binding subject APIGroup must be 'policy.open-cluster-management.io'"
```
Comment (sebrandon1) [inline pkg/ocm/placementbinding.go:276]:
```suggestion
		return "placement binding subject APIGroup must be 'policy.open-cluster-management.io'"
```

File: pkg/ocm/placementbinding.go [hunk 5]
```diff
@@ -180,6 +222,75 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
...
+	}
+
+	if subject.Kind != "Policy" && subject.Kind != "PolicySet" {
+		glog.V(100).Info("The Kind of the subject of the PlacementBinding is invalid")
+
+		errorMsg = "placement binding subject kind must be a valid option"
```
Comment (sebrandon1) [inline pkg/ocm/placementbinding.go:282]:
```suggestion
		return "placement binding subject kind must be a valid option"
```

File: pkg/ocm/placementbinding.go [hunk 6]
```diff
@@ -180,6 +222,75 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
...
+	}
+
+	if subject.Name == "" {
+		glog.V(100).Info("The Name of the subject of the PlacementBinding is empty")
+
+		errorMsg = "placement binding subject name cannot be empty"
```
Comment (sebrandon1) [inline pkg/ocm/placementbinding.go:288]:
```suggestion
		return "placement binding subject name cannot be empty"
```

File: pkg/ocm/placementbinding.go [hunk 7]
```diff
@@ -180,6 +222,75 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
...
+		errorMsg = "placement binding 'PlacementRef.Kind' must be a valid option"
+	}
+
+	if placementRef.Name == "" {
+		glog.V(100).Info("The Name of the PlacementRef of the PlacementBinding is empty")
+
+		errorMsg = "placement binding 'PlacementRef.Name' cannot be empty"
+	}
+
+	return errorMsg
```
Comment (sebrandon1) [inline pkg/ocm/placementbinding.go:265]:
```suggestion
	return ""
```

File: pkg/ocm/placementbinding.go [hunk 8]
```diff
@@ -180,6 +222,75 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
...
+		errorMsg = "placement binding subject kind must be a valid option"
+	}
+
+	if subject.Name == "" {
+		glog.V(100).Info("The Name of the subject of the PlacementBinding is empty")
+
+		errorMsg = "placement binding subject name cannot be empty"
+	}
+
+	return errorMsg
```
Comment (sebrandon1) [inline pkg/ocm/placementbinding.go:291]:
```suggestion
	return ""
```

**Review by sebrandon1 — COMMENTED** [review]
Looks good, just some minor comments about returns.

**Comment by klaskosk** [comment]
@sebrandon1 Thanks for the comments, I've gone ahead and made the changes

**Review by sebrandon1 — APPROVED** [review]

File: pkg/ocm/placementbinding.go [hunk 9]
```diff
@@ -25,6 +25,48 @@ type PlacementBindingBuilder struct {
 	errorMsg string
 }
 
+// NewPlacementBindingBuilder creates a new instance of PlacementBindingBuilder.
+func NewPlacementBindingBuilder(
```
Comment (kononovn) [inline pkg/ocm/placementbinding.go:29]: Are all of these param required for basic resource?

File: pkg/ocm/placementbinding.go [hunk 10]
```diff
@@ -180,6 +222,71 @@ func (builder *PlacementBindingBuilder) Update(force bool) (*PlacementBindingBui
 	return builder, err
 }
 
+// WithAdditionalSubject appends a subject to the subjects list in the PlacementBinding definition.
+func (builder *PlacementBindingBuilder) WithAdditionalSubject(subject policiesv1.Subject) *PlacementBindingBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if err := validateSubject(&subject); err != "" {
```
Comment (kononovn) [inline pkg/ocm/placementbinding.go:231]: I would refactor it to:
```
if err := validateSubject(&subject); err != "" {
    builder.errorMsg = err
    return builder
}

builder.Definition.Subjects = append(builder.Definition.Subjects, subject)

return builder
```

File: pkg/ocm/placementbinding_test.go [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+		Kind:     "PolicySet",
+	}
+)
+
+func TestNewPlacementBindingBuilder(t *testing.T) {
+	// No test cases for invalid refs or subjects, those have their own unit tests.
+	testCases := []struct {
+		placementBindingName      string
+		placementBindingNamespace string
+		placementBindingRef       policiesv1.PlacementSubject
```
Comment (kononovn) [inline pkg/ocm/placementbinding_test.go:31]: Can you add unit tests for:
placementBindingRef = nil
placementBindingSubject = nil ?

File: pkg/ocm/placementbinding_test.go [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+		assert.NotNil(t, placementBindingBuilder)
+		assert.Equal(t, testCase.expectedErrorText, placementBindingBuilder.errorMsg)
+		assert.Equal(t, testCase.placementBindingRef, placementBindingBuilder.Definition.PlacementRef)
+		assert.Equal(t, []policiesv1.Subject{testCase.placementBindingSubject}, placementBindingBuilder.Definition.Subjects)
+	}
+}
+
+func TestWithAdditionalSubject(t *testing.T) {
+	testCases := []struct {
+		subject           policiesv1.Subject
```
Comment (kononovn) [inline pkg/ocm/placementbinding_test.go:76]: What if subject is nil?

File: pkg/ocm/placementbinding_test.go [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+			expectedErrorText: "placement binding 'PlacementRef.Kind' must be a valid option",
+		},
+	}
+
+	for _, testCase := range testCases {
+		err := validatePlacementRef(&testCase.ref)
+		assert.Equal(t, testCase.expectedErrorText, err)
+	}
+}
+
```
Comment (kononovn) [inline pkg/ocm/placementbinding_test.go:149]: Can you please add unit tests for the rest of the functions?

**Review by kononovn — CHANGES_REQUESTED** [review]
Overall looks good. Please check few minor comments above

File: pkg/ocm/placementbinding.go (hunk 9 above)
Comment (klaskosk) [reply pkg/ocm/placementbinding.go:29]: Yes, the kubebuilder validation comments require both of them with all their fields also required

File: pkg/ocm/placementbinding_test.go (hunk 11 above)
Comment (klaskosk) [reply pkg/ocm/placementbinding_test.go:31]: Since the structs are passed directly and not as pointers, they won't be nil. Now that you mentioned it though, the validation functions need tests for nil pointers

Comment (klaskosk) [reply pkg/ocm/placementbinding_test.go:35]: I decided instead to make those functions accept a struct instead of a pointer since they're not modifying the object and then nil pointers won't be an issue

**Review by kononovn — APPROVED** [review]
/lgtm

**Review by cdvultur — APPROVED** [review]
/lgtm
